### PR TITLE
feat: add inline timestamps STT output format

### LINKED ISSUE #222: Feature Request: Add timestamps to speech to text from audio file
Hi, I would like it if you can add a timestamp feature to the speech to text feature when using audio files. I regularly download audio from various websites and transcribe them to text using [TurboScribe](https://turboscribe.ai/). I then later use LLM to summarize the audio and go through it quickly. I would like it if similar feature is implemented in  your app so that I can switch my workflow to my local machine. Also having automatic detection of different speakers would be nice if possible. Thanks.

### PR DESCRIPTION
**Pull request by brittain9**
## Summary
Adds a new **Inline Timestamps** text format option for speech-to-text output, allowing timestamps to be embedded directly within transcribed text as an alternative to SRT subtitle format.

Closes #222

## Screenshots
timestamps

timestamp2

Note: the TTS output at the bottom that strips the current timestamp template.
 
## Motivation
When transcribing audio (podcasts, meetings, interviews), I want timestamps inline with text for:
- Easier human readability vs. SRT's rigid block format
- LLM post-processing (summarization, Q&A) where inline context is more natural
- Quick reference without the overhead of subtitle parsing

### New Settings (Settings → Speech to Text)
| Setting | Description |
|---------|-------------|
| **Text format** | New "Inline Timestamps" option |
| **Timestamp template** | Customizable format with `{hh}`, `{mm}`, `{ss}`, `{ms}`, `{text}` tokens |
| **Minimum interval** | Prevents timestamp spam |

**Example output:** `[00:05] Hello world [00:12] This is a test`

### Implementation
- **`text_tools.cpp`**: Core functions for formatting, regex compilation, and stripping timestamps
- **STT Engines**: Integrated into Vosk, Whisper, FasterWhisper, April, and DeepSpeech
- **TTS Integration**: Auto-strips inline timestamps before speaking (seamless text-to-speech of transcribed content)
- **Bug fix**: Corrected text format resetting to "Plain Text" when clearing notepad

### Tests
- Unit tests covering `format_segments_inline`, `compile_inline_timestamp_regex`, `strip_inline_timestamps`
- Edge cases: interval state across batches, complex template formats, whitespace handling

## Testing Done
- [x] Unit tests pass (`text_tools_test`)
- [x] Manual testing with Vosk, April, Whisper engine 
- [x] TTS strips timestamps before speaking
- [x] Settings UI functional with presets and custom templates

### CODE REVIEW

**Comment by mkiol** [comment]
Sorry for late reply. It looks fantastic :) 

I'm a bit busy at the moment and need a few more days to look at the code and test it. Thank you for your understanding.

File: src/text_tools.cpp [hunk 1]
```diff
@@ -848,6 +848,165 @@ std::string segments_to_subrip_text(const std::vector<segment_t>& segments) {
...
+        if (needs_timestamp) {
+            if (!is_first_entry) os << '\n';
+            os << format_segment_inline(seg, safe_tmpl);
+            last_timestamped_t0 = seg.t0;
+            is_first_entry = false;
+        } else {
+            std::string_view text_view = seg.text;
+            if (!text_view.empty()) {
+                if (text_view.front() == ' ') text_view.remove_prefix(1);
+                os << ' ' << text_view;
```
Comment (mkiol) [inline src/text_tools.cpp:920]: Nice use of string_view 👍🏿

File: src/speech_service.cpp [hunk 2]
```diff
@@ -1285,6 +1291,10 @@ QString speech_service::restart_stt_engine(speech_mode_t speech_mode,
                 static_cast<int>(settings::text_format_t::TextFormatRaw),
                 options)));
         config.sub_config = stt_sub_config_from_options(options);
+        config.inline_timestamp_template =
+            settings::instance()->inline_timestamp_template().toStdString();
+        config.inline_timestamp_min_interval =
+            settings::instance()->inline_timestamp_min_interval();
```
Comment (mkiol) [inline src/speech_service.cpp]: It works, but there is a reason why the config options are not taken directly from the settings object, but are instead passed from `dsnote_app` to `stt_engine` via "options" map. The problem is that Speech Note is split into two separate processes under SFOS, which communicate via DBus. `dsnote_app` has a UI and manages the settings, while `speech_service` is a daemon. When the settings change in "app" process, the "service" process does not detect this. To work around this problem, the settings are passed (via DBus) together with the request type in the "options" map object.

My suggestions:

- Move `inline_timestamp_template` and `inline_timestamp_min_interval` to `config.sub_config` (or create a new structure).
- Retrieve values from the options object in `stt_sub_config_from_options` (or in a new helper function).
- Add `inline_timestamp_template` and `inline_timestamp_min_interval` to the options obj in `dsnote_app::transcribe_file` and `dsnote_app::listen_internal`.
- Add `inline_timestamp_template` to the options in `dsnote_app::play_speech_internal` and `dsnote_app::speech_to_file_internal`.
- Do not call `update_inline_timestamp_regex` when the setting changes, but instead call it in `speech_service::tts_play_speech` and `speech_service::tts_speech_to_file` based on `options` obj.

File: src/speech_service.cpp [hunk 3]
```diff
@@ -345,6 +345,10 @@ speech_service::speech_service(QObject *parent)
             emit default_mnt_out_lang_changed();
         },
         Qt::QueuedConnection);
+    connect(
+        settings::instance(), &settings::inline_timestamp_template_changed,
+            this, &speech_service::update_inline_timestamp_regex);
+    update_inline_timestamp_regex();
```
Comment (mkiol) [inline src/speech_service.cpp]: Timestamp regex is only used to remove inline timestamps for TTS. Am I correct?

Do not call update_inline_timestamp_regex when the setting changes, but instead call it in `speech_service::tts_play_speech` and `speech_service::tts_speech_to_file` based on options obj. I added more on this in the comment below.

File: translations/dsnote-en.ts [hunk 4]
```diff
@@ -78,19 +78,19 @@
 <context>
     <name>AddTextDialog</name>
     <message>
-        <location filename="../sfos/qml/AddTextDialog.qml" line="33"/>
```
Comment (mkiol) [inline translations/dsnote-en.ts:81]: If you are not translating anything, please remove all changes to the TS files from this PR. I will update these files before releasing the new version.

**Review by mkiol — COMMENTED** [review]
Thank you very much for your great work. Love it! I am impressed that you managed to understand the entire wiring. I like your programming style and learned a lot during the review :)

PR is almost ready for merging. The only obstacle is options that are transferred between "app" and "service". The "service" should not read settings that can be changed by the user (the only exception is settings that require a restart). For more info on this problem, see the comments.

**Comment by brittain9** [comment]
Thanks for the suggestions. Pushed the changes. Let me know if anything else needs adjusting.

File: desktop/qml/SettingsSttPage.qml [hunk 5]
```diff
@@ -231,6 +231,126 @@ ColumnLayout {
         }
     }
 
+    ColumnLayout {
+        Layout.fillWidth: true
+        spacing: 0
+        visible: _settings.stt_tts_text_format === Settings.TextFormatInlineTimestamp
```
Comment (mkiol) [inline desktop/qml/SettingsSttPage.qml]:
```suggestion
```

It's a bit confusing that the options aren't always visible. I think it's better to always display them.

**Review by mkiol — COMMENTED** [review]
Thanks for the changes. Everything is working great.

Just a few small requests:

- Always show settings option
- Please remove the *.ts files from the PR

**Comment by brittain9** [comment]
Good suggestion. I originally had some other settings and added that to not bloat up the page. All ts files have been reverted :)

**Review by mkiol — APPROVED** [review]
Perfection. Great feature. Thank you very much for your contribution!
